Source/Core/Common/FileUtil.cpp

```diff
@@ -548,177 +398,173 @@ FSTEntry ScanDirectoryTree(std::string directory, bool recursive)
-    closedir(dirp);
-#endif
 
   return parent_entry;
 }
+#endif
 
-// Deletes the given directory and anything under it. Returns true on success.
-bool DeleteDirRecursively(const std::string& directory)
+// Recursive or non-recursive list of files and directories under directory.
+FSTEntry ScanDirectoryTree(std::string directory, bool recursive)
 {
-  DEBUG_LOG_FMT(COMMON, "DeleteDirRecursively: {}", directory);
-  bool success = true;
+  DEBUG_LOG_FMT(COMMON, "{}: directory {}", __func__, directory);
 
+#ifdef ANDROID
+  if (IsPathAndroidContent(directory))
+    return ScanDirectoryTreeAndroidContent(directory, recursive);
+#endif
+
+  auto path_to_physical_name = [](const fs::path& path) {
 #ifdef _WIN32
-  // Find the first file in the directory.
-  WIN32_FIND_DATA ffd;
-  HANDLE hFind = FindFirstFile(UTF8ToTStr(directory + "\\*").c_str(), &ffd);
+    // TODO Ideally this would not be needed - dolphin really should not have code directly mucking
+    // about with directory separators (for host paths - emulated paths may require it) and instead
+    // use fs::path to interact with them.
+    auto wpath = path.wstring();
+    std::replace(wpath.begin(), wpath.end(), L'\\', L'/');
+    return WStringToUTF8(wpath);
+#else
+    return PathToString(path);
+#endif
+  };
 
-  if (hFind == INVALID_HANDLE_VALUE)
-  {
-    FindClose(hFind);
-    return false;
-  }
+  auto dirent_to_fstent = [&](const fs::directory_entry& entry) {
+    return FSTEntry{
+        .isDirectory = entry.is_directory(),
+        .size = entry.is_directory() ? 0 : entry.file_size(),
+        .physicalName = path_to_physical_name(entry.path()),
+        .virtualName = PathToString(entry.path().filename()),
+    };
+  };
 
-  // Windows loop
-  do
-  {
-    const std::string virtualName(TStrToUTF8(ffd.cFileName));
-#else
-  DIR* dirp = opendir(directory.c_str());
-  if (!dirp)
-    return false;
+  auto calc_dir_size = [](FSTEntry* dir) {
+    dir->size += dir->children.size();
+    for (auto& child : dir->children)
+      if (child.isDirectory)
+        dir->size += child.size;
+  };
 
-  // non Windows loop
-  while (dirent* result = readdir(dirp))
-  {
-    const std::string virtualName = result->d_name;
-#endif
+  const auto directory_path = StringToPath(directory);
 
-    // check for "." and ".."
-    if (((virtualName[0] == '.') && (virtualName[1] == '\0')) ||
-        ((virtualName[0] == '.') && (virtualName[1] == '.') && (virtualName[2] == '\0')))
-      continue;
+  FSTEntry parent_entry;
+  parent_entry.physicalName = path_to_physical_name(directory_path);
+  parent_entry.isDirectory = fs::is_directory(directory_path);
+  parent_entry.size = 0;
 
-    std::string newPath = directory + DIR_SEP_CHR + virtualName;
-    if (IsDirectory(newPath))
+  std::error_code error;
+  if (recursive)
+  {
+    int prev_depth = 0;
+    std::stack<FSTEntry*> dir_fsts;
+    dir_fsts.push(&parent_entry);
+    for (auto it = fs::recursive_directory_iterator(directory_path, error);
```

Comment:
This segment comes across as a bit confusing with the stack and the depth comparisons.  Am I right in assuming we are just trying to come up with a tree of `FSTEntry`s?  And that we can't rely on `recursive_directory_iterator` alone because its order is undefined?

If so, we could improve things from a readability standpoint (not sure if it's more expensive) by avoiding it altogether:

```
void RecurseTreeHelper(auto& fst_entry, auto fs_path)
{
	for (auto it = fs::directory_iterator(fs_path, error); it != fs::directory_iterator();
	 it.increment(error))
	{
		auto child_entry = dirent_to_fstent(*it);
		RecurseTreeHelper(child_entry, *it);
		fst_entry.children.emplace_back(std::move(child_entry));
	}
}
```

Could also do this with a lambda but it's a little klunky:  https://stackoverflow.com/questions/2067988/recursive-lambda-functions-in-c11

If I'm completely missing the point, let me know.

Reply:
Yeah, I didn't take into account the dir size.  Should be easy enough to fix but I get your points.  Out of curiosity, do you avoid recursion due to concerns about stack size or just not liking it?  (I've seen both)

FWIW, I do see what you're doing and it seems like it should work but it isn't intuitive enough to me that I can readily sign off on it.  Unfortunately, we only have unit tests for deletion otherwise I'd have a bit more confidence.  But yeah this is pretty minor, the code overall looks really good.  Great work!